Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy.swift

```diff
@@ -98,24 +102,37 @@ struct PathHierarchy {
                 moduleNode = newModuleNode
                 allNodes[moduleName] = [moduleNode]
             }
+            if let language = language {
+                moduleNode.languages.insert(language)
+            }
             
             var nodes: [String: Node] = [:]
             nodes.reserveCapacity(graph.symbols.count)
             for (id, symbol) in graph.symbols {
                 if let existingNode = allNodes[id]?.first(where: {
                     // If both identifiers are in the same language, they are the same symbol
                     $0.symbol!.identifier.interfaceLanguage == symbol.identifier.interfaceLanguage
-                    // Otherwise, if both have the same name and kind their differences doesn't matter for link resolution purposes
-                    || ($0.name == symbol.pathComponents.last && $0.symbol!.kind.identifier == symbol.kind.identifier)
+                    // Otherwise, if both have the same path components and kind their differences doesn't matter for link resolution purposes
+                    || ($0.symbol!.pathComponents == symbol.pathComponents && $0.symbol!.kind.identifier == symbol.kind.identifier)
                 }) {
                     nodes[id] = existingNode
+                    existingNode.languages.insert(language!) // If we have symbols in this graph we have a language as well
                 } else {
                     assert(!symbol.pathComponents.isEmpty, "A symbol should have at least its own name in its path components.")
                     let node = Node(symbol: symbol, name: symbol.pathComponents.last!)
                     // Disfavor synthesized symbols when they collide with other symbol with the same path.
                     // FIXME: Get information about synthesized symbols from SymbolKit https://github.com/apple/swift-docc-symbolkit/issues/58
-                    node.isDisfavoredInCollision = symbol.identifier.precise.contains("::SYNTHESIZED::")
+                    if symbol.identifier.precise.contains("::SYNTHESIZED::") {
+                        node.specialBehaviors = [.disfavorInLinkCollision, .excludeFromAutomaticCuration]
```

Comment:
The distinction here was because of [my earlier confusion](https://github.com/apple/swift-docc/pull/854#discussion_r1523810534) about why disfavored nodes were being removed from automatic curation. It's a bit redundant now, but with extensions currently being planned for DocC (e.g. [my in-progress PR to create "overload group" symbols](https://github.com/apple/swift-docc/pull/846)) more symbols will become disfavored over time, requiring the assumption to be made explicit.